contracts/AMMStrategy.sol

```diff
@@ -0,0 +1,115 @@
+        for (uint256 i = 0; i < _ammAddrs.length; ++i) {
+            ammMapping[_ammAddrs[i]] = true;
+            emit SetAMM(_ammAddrs[i], true);
+        }
+    }
+
+    /************************************************************
+     *                 Internal function modifier                *
+     *************************************************************/
+    modifier onlyEntryPoint() {
+        require(msg.sender == entryPoint, "only entry point");
+        _;
+    }
+
+    /************************************************************
+     *           Management functions for Owner               *
+     *************************************************************/
+    /// @inheritdoc IAMMStrategy
+    function setEntryPoint(address _newEntryPoint) external override onlyOwner {
+        entryPoint = _newEntryPoint;
+        emit SetEntryPoint(_newEntryPoint);
+    }
+
+    /// @inheritdoc IAMMStrategy
+    function setAMMs(address[] calldata _ammAddrs, bool[] calldata _enables) external override onlyOwner {
+        for (uint256 i = 0; i < _ammAddrs.length; ++i) {
+            ammMapping[_ammAddrs[i]] = _enables[i];
+            emit SetAMM(_ammAddrs[i], true);
+        }
+    }
+
+    /// @inheritdoc IAMMStrategy
+    function approveAssets(
+        address[] calldata _assetAddrs,
+        address[] calldata _ammAddrs,
+        uint256 _assetAmount
+    ) external override onlyOwner {
+        for (uint256 i = 0; i < _assetAddrs.length; ++i) {
+            for (uint256 j = 0; j < _ammAddrs.length; ++j) {
+                IERC20(_assetAddrs[i]).safeApprove(_ammAddrs[j], _assetAmount);
+            }
+        }
+    }
+
+    /************************************************************
+     *                   External functions                      *
+     *************************************************************/
+    /// @inheritdoc IStrategy
+    function executeStrategy(
+        address inputToken,
+        address outputToken,
+        uint256 inputAmount,
+        bytes calldata data
+    ) external override nonReentrant onlyEntryPoint {
+        Operation[] memory ops = abi.decode(data, (Operation[]));
+        require(ops.length > 0, "empty operations");
+        uint256 balanceBefore = IERC20(outputToken).balanceOf(entryPoint);
+        address[] memory opDests = new address[](ops.length);
+        for (uint256 i = 0; i < ops.length; ++i) {
+            Operation memory op = ops[i];
+            require(ammMapping[op.dest], "invalid op dest");
+            opDests[i] = op.dest;
+            _call(op.dest, 0, op.data);
+        }
+        uint256 receivedAmount = IERC20(outputToken).balanceOf(address(this));
+        if (receivedAmount != 0) {
+            IERC20(outputToken).safeTransfer(entryPoint, receivedAmount);
+        }
+        uint256 balanceAfter = IERC20(outputToken).balanceOf(entryPoint);
+        emit Swapped(inputToken, inputAmount, opDests, outputToken, balanceAfter.sub(balanceBefore));
```

Comment:
I think maybe we could have a event for each operation, like cowswap

https://github.com/cowprotocol/contracts/blob/7efd5b43fbf46fe40bac9e6de1b6bab05d27c20b/src/contracts/GPv2Settlement.sol#L462-L466

This can log more info for each operation, and `opDests` could be removed in the swapped event.